---
title: "Differential Transcript Expression of 42 vs 44 colorectal cancer samples in R"
author: "Salomón Marquez"
date: "`r Sys.Date()`"
output:
   html_document:
    code_folding: hide
    toc: true
    toc_float:
      collapsed: true
      smooth_scroll: true
    toc_depth: 3
    number_sections: true
    self-contained: true
#output:
#   pdf_document:
#   latex_engine: xelatex  # o lualatex
#   keep_tex: yes
#   toc: yes
editor_options: 
  markdown: 
    wrap: 72
#editor_options: 
#  chunk_output_type: console
---


```{r setup, include=FALSE}
# Load knitr
library(knitr)

# Global R options
options(max.print = 50)
options(repos = c(CRAN = "https://cloud.r-project.org")) 

# Global chunk options
opts_chunk$set(
  echo = TRUE,     # show code
  cache = TRUE,    # caching
  prompt = FALSE,   # don’t show ">" in code
  tidy = TRUE,      # clean code formatting
  comment = NA,     # no extra comment symbols
  message = FALSE,  # hide package load messages
  warning = FALSE   # hide warnings
)

# Knitr rendering width
opts_knit$set(width = 100)
```

```{r, include=FALSE}
# Install and load modules and libraries 
library(skimr)
library(scales)
library(ggplot2)
library(tidyr)
library(dplyr)
library(S4Vectors)
library(reshape2)
library(ggfortify)
library(corrplot)
library(factoextra)

library(SummarizedExperiment)
library(DESeq2)
library(IsoformSwitchAnalyzeR)
library(tidyverse)
library(dbplyr)
library(BiocParallel)
library(Biostrings)
```

## Load dataset
Let's ensure that the previously mentioned file are in our working directory. 
```{r}
# Specify a working directory
directory <- "/Volumes/DataWinBackup/00_MASTER UOC BIOINFORMATICA - TEMPORAL/TFM-UOC/DTU44CRC"
setwd(directory)
list.files()
```
```{r}
# Load file
mySwitchList <- readRDS("mySwitchList_consequences_42_44.rds")

# [1] "aSwitchList" "counts"      "directory"   "myDesign"    "tpm"  
```

```{r}
# View salmonQuant
names(mySwitchList)
head(mySwitchList, 3) 
```

```{r}
# Summary
extractSwitchSummary(mySwitchList, filterForConsequences = F)
```

```{r}
# Genome-wide Summaries
extractSwitchOverlap(
    mySwitchList,
    filterForConsequences=TRUE
)
```

```{r}
# Extract 10 top switching genes (by q-value)
extractTopSwitches(
    mySwitchList, 
    filterForConsequences = FALSE, 
    n = 10, 
    sortByQvals = TRUE
)
```

## Analysis of Individual Isoform Switching
```{r}
# Visualize gene transcripts variants
switchPlotTranscript(mySwitchList, gene="TPM2") # Visualizes the transcripts and their annotation
```
```{r}
# Visualizes the gene expression
switchPlotGeneExp(mySwitchList, gene="TPM2", condition1 = 'normal_eocrc',condition2 = 'tumor_eocrc', localTheme = theme_bw(base_size = 9)) 
```
```{r}
# Visualizes the gene expression
switchPlotIsoExp(mySwitchList, gene="TPM2", condition1 = 'normal_eocrc',condition2 = 'tumor_eocrc', localTheme = theme_bw(base_size = 9)) 
```
```{r}
# Visualizes the isoform usage 
switchPlotIsoUsage(mySwitchList, gene="TPM2", condition1 = 'normal_eocrc',condition2 = 'tumor_eocrc', localTheme = theme_bw(base_size = 9)) 
```
```{r}
# Summary
switchPlot(mySwitchList, gene="TPM2", condition1 = 'normal_eocrc',condition2 = 'tumor_eocrc', localTheme = theme_bw(base_size = 9))  
```
## Genome-wide Summaries 

```{r}
# Genome-Wide Analysis of Alternative Splicing
extractSplicingSummary(
    mySwitchList,
    returnResult = FALSE # if TRUE returns a data.frame with the summary statistics
)
```

```{r}
# Genome-wide Summaries
extractConsequenceSummary(
    mySwitchList,
    consequencesToAnalyze="all",
    plotGenes = FALSE,           # enables analysis of genes (instead of isoforms)
    asFractionTotal = FALSE      # enables analysis of fraction of significant features
)
```
## Consequence Enrichment Analysis 
```{r}
extractConsequenceEnrichment(
    mySwitchList,
    consequencesToAnalyze='all',
    analysisOppositeConsequence = TRUE,
    localTheme = theme_bw(base_size = 14), 
    returnResult = FALSE 
)
```
## Splicing Enrichment Analysis
```{r}
extractSplicingEnrichment(
    mySwitchList,
    returnResult = FALSE # if TRUE returns a data.frame with the summary statistics
)
```

```{r}
extractSplicingEnrichmentComparison(
    mySwitchList,
    splicingToAnalyze = c('A3','A5','ATSS','ATTS'), # the splicing highlighted above
    returnResult = FALSE # if TRUE returns a data.frame with the results
)
```


```{r}
extractSplicingGenomeWide(
    mySwitchList,
    returnResult = FALSE # if TRUE returns a data.frame with the summary statistics
)
```





## Overview Plots
```{r}
# Volcano like plot:
ggplot(data=mySwitchList$isoformFeatures, aes(x=dIF, y=-log10(isoform_switch_q_value))) +
     geom_point(
        aes( color=abs(dIF) > 0.1 & isoform_switch_q_value < 0.05 ), # default cutoff
        size=1
    ) +
    geom_hline(yintercept = -log10(0.05), linetype='dashed') + # default cutoff
    geom_vline(xintercept = c(-0.1, 0.1), linetype='dashed') + # default cutoff
    facet_wrap( ~ condition_2) +
    #facet_grid(condition_1 ~ condition_2) + # alternative to facet_wrap if you have overlapping conditions
    scale_color_manual('Signficant\nIsoform Switch', values = c('black','red')) +
    labs(x='dIF', y='-Log10 ( Isoform Switch Q Value )') +
    theme_bw()
```
```{r}
# Switch vs Gene changes:
ggplot(data=mySwitchList$isoformFeatures, aes(x=gene_log2_fold_change, y=dIF)) +
    geom_point(
        aes( color=abs(dIF) > 0.1 & isoform_switch_q_value < 0.05 ), # default cutoff
        size=1
    ) + 
    facet_wrap(~ condition_2) +
    #facet_grid(condition_1 ~ condition_2) + # alternative to facet_wrap if you have overlapping conditions
    geom_hline(yintercept = 0, linetype='dashed') +
    geom_vline(xintercept = 0, linetype='dashed') +
    scale_color_manual('Signficant\nIsoform Switch', values = c('black','red')) +
    labs(x='Gene log2 fold change', y='dIF') +
    theme_bw()
```